pkg/util/containers/providers/windows/provider.go

```diff
@@ -72,38 +72,66 @@ func (mp *provider) Prefetch() error {
 		return err
 	}
 
+	log.Debugf("Retrieved %d containers from docker", len(rawContainers))
+
 	// Used to find if Agent is running in a container.
 	// With K8S entrypoint, `agentPID` should match
 	// With Docker entrypoint, `parentPID` should match
 	agentPID := os.Getpid()
 	parentPID := os.Getppid()
 
 	containers := make(map[string]containerBundle, len(rawContainers))
-	for _, container := range rawContainers {
-		containerBundle := containerBundle{}
-
-		cjson, err := dockerUtil.Inspect(context.TODO(), container.ID, false)
-		if err == nil {
-			mp.fillContainerDetails(cjson, &containerBundle)
-
-			// Luckily for us, on Windows PIDs are the same inside/outside containers
-			if cjson.State.Pid == agentPID || cjson.State.Pid == parentPID {
-				mp.agentCID = &container.ID
+	var containersLock = sync.RWMutex{}
+	var wg sync.WaitGroup
+	// On Windows fetching the info on docker containers can be slow.
+	// On a host with ~100 hosts running, this can easily take up more than 30s,
+	// causing the Agent to appear 'stuck' and the entrypoint/SCM to consider the Agent dead.
+	// Divide the fetch into batches to accelerate this process; here 8 is chosen arbitrarily.
+	chunkSize := len(rawContainers) / 8
+	if chunkSize <= 1 {
+		chunkSize = len(rawContainers)
+	}
+	log.Infof("Fetching container info by batch of %d\n", chunkSize)
+	for i := 0; i < len(rawContainers); i += chunkSize {
```

Comment:
I do not think I understand the logic of chunking. If I have 16 containers how many chunks it should be. I thought it should be 2 by 8 where the 8 is a batch size. Yet in iteration you increment not by batch size but by chunk size which will cause creating 8 batches of size 2. Maybe I am missing something

Reply:
Nice idea though